tls/s2n_recv.c

```diff
@@ -228,7 +229,9 @@ ssize_t s2n_recv_impl(struct s2n_connection *conn, void *buf, ssize_t size_signe
         }
     }
 
-    if (s2n_stuffer_data_available(&conn->in) == 0) {
+    /* Clear the blocked status if the stuffer was completely drained or we returned at */
+    /* least one byte to the application. */
+    if (s2n_stuffer_data_available(&conn->in) == 0 || bytes_read > 0) {
```

Comment:
I was mostly trying to preserve an already-fragile interface contract. I initially removed it but had a test fail and thought that was good enough justification for keeping it :smile:

Reply:
Looking at the failure it's coming from the `s2n_alerts_protocol_test`:

```c
FAILED test 62297
(blocked) == (S2N_NOT_BLOCKED) is not true  (~/s2n-tls/tests/unit/s2n_alerts_protocol_test.c:318)
Error Message: 'no error'
 Debug String: '(null)'
 System Error: Success (0)
Running ~/s2n-tls/tests/unit/s2n_alerts_protocol_test.c ...
```

https://github.com/aws/s2n-tls/blob/05ae36ce470f20233821f3cb000140124023b6bb/tests/unit/s2n_alerts_protocol_test.c#L316-L318

Reply:
After taking with @goatgoose, I don't think we can make the `bytes_read > 0` change. Historically, apparently we've used the blocked status to signal whether or not we drained the stuffer. It will likely break someone out there and is actually referred to several times in our documentation (which needs to be fixed separately) :frowning_face: 

https://github.com/aws/s2n-tls/blob/2da0bf416cae49bc2a58222494c1b3ef5aa7f492/api/s2n.h#L1852-L1861

https://github.com/aws/s2n-tls/blob/2da0bf416cae49bc2a58222494c1b3ef5aa7f492/docs/USAGE-GUIDE.md?plain=1#L335-L336